### PR TITLE
Expand Python Bindings to include S2RegionTermIndexer

### LINKED ISSUE #84: Python bindings do not include S2RegionTermIndexer
`S2RegionTermIndexer` contains useful logic that would be great to have in the Python bindings.

Is there a particular reason this class isn't included? I'm unfortunately out of my depth in trying to modify the bindings, and instead planning to just reimplement some of it in the python code directly. Thanks.

### PR DESCRIPTION
**Pull request by rcoup**
* Make it possible to follow the _Indexing for Search_ and _Querying for Search_ sections in the [Quickstart guide](https://s2geometry.io/devguide/cpp/quickstart#indexing-for-search) from Python.
* Primarily expand the Python bindings to include `S2RegionTermIndexer`
* And related changes for testing
* Port the `term_index.cc` example to Python as well for good measure.
* Add some `repr()` methods

Fixes #84

### CODE REVIEW

**Comment by rcoup** [comment]
@googlebot I signed it!

File: src/python/pywraps2_test.py [hunk 1]
```diff
@@ -606,5 +632,166 @@ def testS2CellIdToFromFaceIJ(self):
...
+        if covering.Intersects(coverings[j]):
+          expected.add(j)
+      
+      for term in terms:
+        actual |= index[term]
+
+      self.assertEqual(expected, actual)
+      query_terms += len(terms)
+
+      # print("Index terms/doc: %0.2f, Query terms/doc: %0.2f" % (float(index_terms) / ITERATIONS, float(query_terms) / ITERATIONS))
```
Comment (jmr) [inline src/python/pywraps2_test.py:718]: might as well print it since the C++ version does

File: src/python/s2_common.i [hunk 2]
```diff
@@ -34,6 +36,16 @@
 %apply std::vector<S2CellId> *OUTPUT {std::vector<S2CellId> *interior};
 %apply std::vector<S2CellId> *OUTPUT {std::vector<S2CellId> *output};
 
+%typemap(in) absl::string_view {
+  if (PyUnicode_Check($input)) {
+    $1 = absl::string_view(PyUnicode_AsUTF8($input));
+  } else {
+    SWIG_exception(SWIG_TypeError, "string expected");
```
Comment (jmr) [inline src/python/s2_common.i:43]: a string_view may also be pointing at non-utf-8 data, but I guess that's not happening. How do we add that if we need to?

File: src/python/s2_common.i [hunk 3]
```diff
@@ -127,8 +139,17 @@ std::vector<S2Polyline *> *out(std::vector<S2Polyline *> temp) {
   }
 }
 
+%inline %{
+  // mostly useful for testing
+  static PyObject *S2Point_FromRaw(double x, double y, double z) {
```
Comment (jmr) [inline src/python/s2_common.i:144]: Can this be a constructor?

File: src/python/s2_common.i [hunk 4]
```diff
@@ -127,8 +139,17 @@ std::vector<S2Polyline *> *out(std::vector<S2Polyline *> temp) {
...
+    return SWIG_NewPointerObj(new S2Point(x, y, z), SWIGTYPE_p_S2Point,
+                              SWIG_POINTER_OWN);
+  }
+%}
+
 // We provide our own definition of S2Point, because the real one is too
 // difficult to wrap correctly.
+
```
Comment (jmr) [inline src/python/s2_common.i:152]: delete this added empty line

File: src/python/s2_common.i [hunk 5]
```diff
@@ -266,6 +334,8 @@ class S2Point {
 %unignore S2Cap::Empty;
 %unignore S2Cap::Encode;
 %unignore S2Cap::Expanded;
+%unignore S2Cap::FromCenterHeight(const S2Point&, double);
+%unignore S2Cap::FromCenterArea(const S2Point&, double);
```
Comment (jmr) [inline src/python/s2_common.i:338]: Keep these in alphabetical order.

File: src/python/s2_common.i [hunk 6]
```diff
@@ -362,6 +433,41 @@ class S2Point {
 %unignore S2CellUnion::cell_ids;
 %unignore S2CellUnion::empty;
 %unignore S2CellUnion::num_cells;
+%unignore S2Earth;
```
Comment (jmr) [inline src/python/s2_common.i:436]: alphabetical order, use your editor to sort them

**Review by jmr — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: doc/examples/term_index.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+This is a port of the C++ term_index.cc example for the Python API.
+"""
+import argparse
+import random
+from collections import defaultdict
+
+import pywraps2 as s2
+
+
+def RandomPoint():
```
[the remark this answers is not in the document]
Comment (rcoup) [reply doc/examples/term_index.py:20]: The S2Testing C++ class is conditionally built, and not with #defines, so it didn't seem trivial to conditionally build Python bindings to it. For the time I had, seemed easier to reimplement it.

File: src/python/s2_common.i (hunk 2 above)
Comment (rcoup) [reply src/python/s2_common.i:43]: The string in this case is coming from python though, and [SWIG's default behaviour is to map std::string ↔︎ String](http://www.swig.org/Doc3.0/Python.html#Python_nn77) in Py3.

Could use `SWIG_PYTHON_STRICT_BYTE_CHAR` and force all string encoding to happen in Python. Not sure if that solves a real use-case though?

File: src/python/s2_common.i (hunk 3 above)
Comment (rcoup) [reply src/python/s2_common.i:144]: I had a go at that initially and ran into issues because of the [`S2Point` redefinition](https://github.com/google/s2geometry/blob/master/src/python/s2_common.i#L130-L137) happening. Suggestions welcome.

File: src/python/pywraps2_test.py (hunk 1 above)
Comment (rcoup) [reply src/python/pywraps2_test.py:718]: unittest has terrible output-control cf pytest/etc, but I guess can suppress via `-b`

File: src/python/s2_common.i (hunk 2 above)
Comment (rcoup) [reply src/python/s2_common.i:43]: I guess I could easily convert Python Bytes without encoding to `std::string`, but not sure what would happen on the way back out, considering SWIG's default behaviour.

File: src/python/pywraps2_test.py [hunk 8]
```diff
@@ -14,12 +14,38 @@
...
-
-
+import random
 import unittest
+from collections import defaultdict
+
 import pywraps2 as s2
 
 
+class S2Testing(object):
```
[the remark this answers is not in the document]
Comment (rcoup) [reply src/python/pywraps2_test.py]: I moved it to `pywraps2.S2Testing`

**Comment by rcoup** [comment]
@jmr ready for another look

File: src/python/s2_common.i (hunk 3 above)
Comment (jmr) [reply src/python/s2_common.i:144]: Could you just add a comment that that's what's happening? Maybe we'll figure out how to clean it up later.

Comment (rcoup) [reply src/python/s2_common.i:145]: done

**Comment by rcoup** [comment]
@googlebot I fixed it

**Comment by rcoup** [comment]
@jmr I think this is good to go now.

File: src/python/s2_common.i [hunk 9]
```diff
@@ -597,3 +723,72 @@ USE_COMPARISON_FOR_LT_AND_GT(S1Angle)
 USE_EQUALS_FN_FOR_EQ_AND_NE(S2Loop)
 USE_EQUALS_FN_FOR_EQ_AND_NE(S2Polygon)
 USE_EQUALS_FN_FOR_EQ_AND_NE(S2Polyline)
+
```
Comment (jmr) [inline src/python/s2_common.i:726]: Can this section be a separate PR? Everything else looks good, but now I have to go learn about repr()

File: src/python/s2_common.i [hunk 10]
```diff
@@ -597,3 +723,72 @@ USE_COMPARISON_FOR_LT_AND_GT(S1Angle)
...
+// repr() methods for geodetic/cell classes
+
+%extend S2CellId {
+  %pythoncode %{
+    def __repr__(self):
+      return '<S2CellId: %s>' % self.ToToken()
```
Comment (jmr) [inline src/python/s2_common.i:732]: Should this eval back to the same thing? If not, ToString() is more readable and standard for debugging in c++.

**Comment by rcoup** [comment]
Done, will do a new PR for repr
